### PR TITLE
Use Identity/Reference based collection where possible

### PR DESCRIPTION
**Pull request by ImMorpheus**
[SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/3132) | **SpongeCommon**

### CODE REVIEW

**Comment by dualspiral** [comment]
I'm curious, what are the performance gains from doing this?

**Comment by ImMorpheus** [comment]
https://docs.oracle.com/javase/8/docs/api/java/util/IdentityHashMap.html

> For many JRE implementations and operation mixes, this class will yield better performance than HashMap (which uses chaining rather than linear-probing). 

We're already largely taking advantage of the performance gain of `IdentityHashMap` in the codebase, this is also a matter of consistency. Some of the collections were not using it even if they could. 
The performance gain is hard to measure because this PR is not targeting hotspots:
Ex.
`ChunkSaveHelper#writeChunks` is only used by the command `/sponge chunks`
the `VirtualPortalAgent` class is not used at all

the `builders` field in `SpongeDataManager` on the other hand is used when serializing/deserializing `MemoryDataContainer`s

**Comment by Zidane** [comment]
Speaking of inconsistency...

Can we stop using Guava to create basic maps like this? I see `Sets.` and then `new Identity`.

**Comment by ImMorpheus** [comment]
> Speaking of inconsistency...
> 
> Can we stop using Guava to create basic maps like this? I see `Sets.` and then `new Identity`.

While I hate the abuse of Guava, there's nothing I can do in this case.
Java doesn't have an `IdentityHashSet`.

EDIT: as a matter of fact `new` is used with Identity**HashMap**s
EDIT2: k, moved to fastutil. Now there's "new" everywhere.

**Comment by Zidane** [comment]
Ha @ImMorpheus I completely forgot about that...

Ignore me.

**Comment by Zidane** [comment]
@ImMorpheus Remove the bad Sets imports, I think they are left over before you replaced Guava usages.

Otherwise, great PR. Thanks as always.

File: src/main/java/org/spongepowered/common/event/tracking/phase/packet/inventory/BasicInventoryPacketState.java [hunk 1]
```diff
@@ -24,6 +24,8 @@
  */
 package org.spongepowered.common.event.tracking.phase.packet.inventory;
 
+import com.google.common.collect.Sets;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/event/tracking/phase/packet/inventory/BasicInventoryPacketState.java]:
```suggestion
```

File: src/main/java/org/spongepowered/common/item/inventory/archetype/SpongeInventoryArchetypeBuilder.java [hunk 2]
```diff
@@ -24,6 +24,8 @@
  */
 package org.spongepowered.common.item.inventory.archetype;
 
+import com.google.common.collect.Sets;
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/SpongeInventoryArchetypeBuilder.java]:
```suggestion
```

File: src/main/java/org/spongepowered/common/item/inventory/lens/impl/ReusableLens.java [hunk 3]
```diff
@@ -65,7 +66,7 @@ private ReusableLens(SlotProvider slots, Function<SlotProvider, T> lensProvider)
     public static <T extends Lens> ReusableLens<T> getLens(Class<T> lensType, InventoryAdapter adapter,
             Supplier<SlotProvider> slots, Function<SlotProvider, T> lens) {
         Map<Class<? extends Lens>, Int2ObjectMap<ReusableLens>>
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/inventory/lens/impl/ReusableLens.java:67]: Isn't this going to be changing to a different type of map?

**Comment by ImMorpheus** [comment]
yep, I'll remove the old imports with the next commits. 

I've also found out that `Key`s and `EventContextKey`s are treated as if they don't have an equals contract (compared with `==` and used in `IdentityHashMap`s) so there are a few spots that can be updated.

**Comment by ImMorpheus** [comment]
Ok, this is ready now.
Turned out `EvenContextKey`s are following the `equals` contract, so unless we decide that it's perfectly fine to compare them with `==` or use them in Identity maps then they should stay as they are now.
`Key`s are already being used that way (see https://github.com/SpongePowered/SpongeCommon/blob/a4f4c117ad89266a57762eaf27cd01d30251a7bb/src/main/java/org/spongepowered/common/data/SpongeManipulatorRegistry.java#L94 for an example)

The map passed to `NotifyNeighbourEvent` is now an EnumMap, that should help a bit given how often that's called.

File: src/main/java/org/spongepowered/common/data/property/SpongePropertyRegistry.java [hunk 4]
```diff
@@ -104,7 +105,7 @@ public void completeRegistration() {
     public Collection<Property<?, ?>> getPropertiesFor(final PropertyHolder holder) {
         final ImmutableList.Builder<Property<?, ?>> builder = ImmutableList.builder();
         if (this.tempRegistry != null) { // Still doing registrations
-            final Set<Class<? extends Property<?, ?>>> used = new HashSet<>();
+            final Set<Class<? extends Property<?, ?>>> used = new ReferenceOpenHashSet<>();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/property/SpongePropertyRegistry.java]: this is a moot set to even bother using. It's a temp registry that gets dropped as soon as the game has finished loading.

File: src/main/java/org/spongepowered/common/item/inventory/lens/impl/ReusableLens.java (hunk 3 above)
Comment (ImMorpheus) [reply src/main/java/org/spongepowered/common/item/inventory/lens/impl/ReusableLens.java:67]: `IdentityHashMap` implements `Map`. The declaration is fine, I'm changing the value here
https://github.com/SpongePowered/SpongeCommon/pull/2569/files#diff-79c49477a2574946c4d8b9f22e05e208R68

File: src/main/java/org/spongepowered/common/data/util/DataUtil.java [hunk 5]
```diff
@@ -580,10 +581,9 @@ public static void readCustomData(final NBTTagCompound compound, final DataHolde
                 // Re-attempt to deserialize custom data
                 final SerializedDataTransaction transaction = deserializeManipulatorList(builder.build());
                 final List<DataManipulator<?, ?>> manipulators = transaction.deserializedManipulators;
-                final List<Class<? extends DataManipulator<?, ?>>> classesLoaded = new ArrayList<>();
+                final Set<Class<? extends DataManipulator<?, ?>>> classesLoaded = new ReferenceOpenHashSet<>();
                 for (final DataManipulator<?, ?> manipulator : manipulators) {
-                    if (!classesLoaded.contains(manipulator.getClass())) {
-                        classesLoaded.add((Class<? extends DataManipulator<?, ?>>) manipulator.getClass());
+                    if (classesLoaded.add((Class<? extends DataManipulator<?, ?>>) manipulator.getClass())) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/util/DataUtil.java:586]: Damn, this must've been here for ages.....

**Review by gabizou — COMMENTED** [review]
looks good. Would be good to get some profiling results if you can, Might be worth spinning up some stupid Redstone machine that uses hoppers and enable `-Dsponge.shouldfireAll=true` to ping the event listeners and phase states and see what the comparison is. That being said, I'd like to take that approach to do some base testing with some profiling results as an example (acknowledging some limited testing, and of course, letting the JVM "warm up").

File: src/main/java/org/spongepowered/common/data/property/SpongePropertyRegistry.java (hunk 4 above)
Comment (ImMorpheus) [reply src/main/java/org/spongepowered/common/data/property/SpongePropertyRegistry.java]: Reverted with other temporary maps.

**Pull request merged**